server/src/main/java/org/elasticsearch/index/engine/SoftDeletesPolicy.java

```diff
@@ -0,0 +1,114 @@
+     */
+    synchronized void setRetentionOperations(long retentionOperations) {
+        this.retentionOperations = retentionOperations;
+    }
+
+    /**
+     * Sets the local checkpoint of the current safe commit.
+     * All operations whose seqno are greater than this checkpoint will be retained until the new checkpoint is advanced.
+     */
+    synchronized void setCheckpointOfSafeCommit(long newCheckpoint) {
+        if (newCheckpoint < this.checkpointOfSafeCommit) {
+            throw new IllegalArgumentException("Local checkpoint can't go backwards; " +
+                "new checkpoint [" + newCheckpoint + "]," + "current checkpoint [" + checkpointOfSafeCommit + "]");
+        }
+        this.checkpointOfSafeCommit = newCheckpoint;
+        updateMinRequiredSeqNoForRecovery();
+    }
+
+    private void updateMinRequiredSeqNoForRecovery() {
+        assert Thread.holdsLock(this) : Thread.currentThread().getName();
+        if (retentionLockCount == 0) {
+            // Need to keep all operations after the local checkpoint of the safe commit for recovery purpose.
+            this.minRequiredSeqNoForRecovery = checkpointOfSafeCommit + 1;
+        }
+    }
+
+    /**
+     * Acquires a lock on soft-deleted documents to prevent them from cleaning up in merge processes. This is necessary to
+     * make sure that all operations after the local checkpoint of the safe commit are retained until the lock is released.
+     * This is a analogy to the translog's retention lock; see {@link Translog#acquireRetentionLock()}
+     */
+    synchronized Releasable acquireRetentionLock() {
+        assert retentionLockCount >= 0 : "Invalid number of retention locks [" + retentionLockCount + "]";
+        retentionLockCount++;
+        final AtomicBoolean released = new AtomicBoolean();
+        return () -> {
+            if (released.compareAndSet(false, true)) {
+                releaseRetentionLock();
+            }
+        };
+    }
+
+    private synchronized void releaseRetentionLock() {
+        assert retentionLockCount > 0 : "Invalid number of retention locks [" + retentionLockCount + "]";
+        retentionLockCount--;
+        updateMinRequiredSeqNoForRecovery();
+    }
+
+    /**
+     * Returns a soft-deletes retention query that will be used in {@link org.apache.lucene.index.SoftDeletesRetentionMergePolicy}
+     * Documents including tombstones are soft-deleted and matched this query will be retained and won't cleaned up by merges.
+     */
+    Query retentionQuery() {
+        return LongPoint.newRangeQuery(SeqNoFieldMapper.NAME, getMinSeqNoToRetain(), Long.MAX_VALUE);
+    }
+
+    // Package-level for testing
+    synchronized long getMinSeqNoToRetain() {
+        final long minSeqNoForQueryingChanges = globalCheckpointSupplier.getAsLong() + 1 - retentionOperations;
+        return Math.min(minRequiredSeqNoForRecovery, minSeqNoForQueryingChanges);
```

Comment:
instead of adding the `1` on assignment which is done in 2 places should we do it here?